qubes/storage/lvm.py

```diff
@@ -263,68 +360,94 @@ def _remove_revisions(self, revisions=None):
-                key=operator.itemgetter(1))
+                key=_revision_sort_key)
             # pylint: disable=invalid-unary-operand-type
             revisions = revisions[:(-self.revisions_to_keep) or None]
             revisions = [rev_id for rev_id, _ in revisions]
 
         for rev_id in revisions:
+            # safety check
+            assert rev_id != self._vid_current
             try:
                 cmd = ['remove', self.vid + '-' + rev_id]
                 qubes_lvm(cmd, self.log)
             except qubes.storage.StoragePoolException:
                 pass
 
-    def _commit(self):
+    def _commit(self, vid_to_commit=None, keep=False):
+        '''
+        Commit temporary volume into current one. By default
+        :py:attr:`_vid_snap` is used (which is created by :py:meth:`start()`),
+        but can be overriden by *vid_to_commit* argument.
+
+        :param vid_to_commit: LVM volume ID to commit into this one
+        :param keep: whether to keep or not *vid_to_commit*.
+          IOW use 'clone' or 'rename' methods.
+        :return: None
+        '''
         msg = "Trying to commit {!s}, but it has save_on_stop == False"
         msg = msg.format(self)
         assert self.save_on_stop, msg
 
         msg = "Trying to commit {!s}, but it has rw == False"
         msg = msg.format(self)
         assert self.rw, msg
-        assert hasattr(self, '_vid_snap')
+        if vid_to_commit is None:
+            assert hasattr(self, '_vid_snap')
+            vid_to_commit = self._vid_snap
 
-        if self.revisions_to_keep > 0:
-            cmd = ['clone', self.vid,
-                '{}-{}-back'.format(self.vid, int(time.time()))]
+        # TODO: when converting this function to coroutine, this _must_ be
+        # under a lock
+        if not os.path.exists('/dev/' + vid_to_commit):
+            # nothing to commit
+            return
+
+        previous_vid_current = self._vid_current
+        if previous_vid_current == self.vid:
+            # VM started with old code, need to rename volume
+            # to contain revision_id
+            cmd = ['rename', previous_vid_current,
+                   '{}-{}.{}'.format(
+                    self.vid, self._get_next_revision_number(),
+                    int(time.time()))]
             qubes_lvm(cmd, self.log)
             reset_cache()
-        self._remove_revisions()
 
-        # TODO: when converting this function to coroutine, this _must_ be
-        # under a lock
-        # remove old volume only after _successful_ clone of the new one
-        cmd = ['rename', self.vid, self.vid + '-tmp']
+        # then assign new revision number to the volume we're committing right
+        # now
+        cmd = ['clone' if keep else 'rename',
+               vid_to_commit,
+               '{}-{}.{}'.format(
+                self.vid, self._get_next_revision_number(),
+                int(time.time()))]
         qubes_lvm(cmd, self.log)
```

Comment:
I don't have an LVM installation at hand (and I'm a bit out of my depth in the `lvm.py` code) - but it looks like the revision is now created directly from the dirty data after shutdown, instead of from the clean data committed during the _previous_ shutdown? Changing the meaning of `revisions_to_keep` (because there's no longer an implicit not-yet-a-revision image), and making `revert()` without an explicit `revision` argument just move the latest data to a "new" revision with a clobbered timestamp.

Reply:
Monotonic revision numbers are a nice touch. I shied away from those when coding the reflink driver, IIRC due to being nervous about leaving an obvious VM startup counter around in case of dom0 compromise. But ordering revisions correctly when the clock has changed seems more pressing than that. Maybe a random initial offset could be added, or something? Anyway, I've ported the monotonic design to reflink.

But it might make sense to keep the current dirty -> clean -> revision data lifecycle, which IMO leads itself to a more natural definition of `revisions_to_keep` and `revert`.

Reply:
This is preserved - see that the most recent is excluded from self.revisions.

Reply:
The whole point of this change is to have commit a single step operation, instead of two (mark previous "most recent" as "old", then dirty -> "most recent").